content/blog/build-publish-containers-iac/index.md

```diff
@@ -0,0 +1,2252 @@
+```
+
+This image name is the versioned one, meaning that any time a new version is created, Pulumi will detect that it needs to update the Kubernetes deployment which references it, triggering a rollout of the new deployed image. As such, we just need to run `pulumi up` anytime the application or its `Dockerfile` changes and Pulumi will detect and trigger a deployment to Kubernetes.
+
+To see this in action, let's change our `Dockerfile`'s contents once more:
+
+```dockerfile
+FROM nginx
+RUN echo "<h1>Hello, World -- from Pulumi and Kubernetes!</h1>" > \
+    /usr/share/nginx/html/index.html
+```
+
+And then rerun Pulumi -- notice that both the image _and_ the Kubernetes deployment has changed this time:
+
+```bash
+$ pulumi up
+Updating (dev)
+
+     Type                              Name            Status      Info
+ +   pulumi:pulumi:Stack               project-dev     created
+ ~   ├─ docker:image:Image             my-image        updated
+ ~   └─ kubernetes:core/v1:Deployment  app-dep         updated     [diff: ~spec]
+
+Outputs:
+    appIp        : "34.82.32.166"
+    baseImageName: "...cloud-specific url.../my-repo-dc811b0"
+    fullImageName: "...cloud-specific url.../my-repo-dc811b0:36919b56f6b2dedfa879815443c9964d17a89506caed25025328b1d0204a083f"
+```
+
+Finally, if we re-curl the endpoint, we will now see that the updated container is now live!
+
+```
+$ curl $(pulumi stack output appIp)
+<h1>Hello, World -- from Pulumi and Kubernetes!</h1>
+```
+
+And there we go: we have gone from a `Dockerfile` to a published container image in a container registry of our choosing, which has been used to spin up a load balanced Kubernetes service, and which is now ready for fully automated continuous deployments.
+
+## Wrapping Up
+
+In this article, we've seen how easy it is to build, publish, and use container images in many popular public and private container registry options. We have seen how to provision new registries using infrastructure as code and how easy it is to trigger deployments of application updates by simply running a single `pulumi up` command.
+
+All of these steps were manually run from a CLI, however, a natural next step is to [wire the entire process up to a CI/CD system]({{< relref "/docs/guides/continuous-delivery" >}}) such as [GitHub Actions]({{< relref "/docs/guides/continuous-delivery/github-actions" >}}), [GitLab Pipelines]({{< relref "/docs/guides/continuous-delivery/gitlab-ci" >}}), [Jenkins]({{< relref "/docs/guides/continuous-delivery/jenkins" >}}), [Spinnaker]({{< relref "/docs/guides/continuous-delivery/spinnaker" >}}), or [one of the many available options]({{< relref "/docs/guides/continuous-delivery" >}}, so that you can deploy continuously as you merge code. For even more advanced scenarios, we might want to build a custom program such as a CLI that uses the [Automation API]({{< relref "/blog/automation-api" >}}) to perform these actions behind a simpler, purpose-built interface.
```

Comment:
```suggestion
All of these steps were manually run from a CLI; however, a natural next step is to [wire the entire process up to a CI/CD system]({{< relref "/docs/guides/continuous-delivery" >}}) such as [GitHub Actions]({{< relref "/docs/guides/continuous-delivery/github-actions" >}}), [GitLab Pipelines]({{< relref "/docs/guides/continuous-delivery/gitlab-ci" >}}), [Jenkins]({{< relref "/docs/guides/continuous-delivery/jenkins" >}}), [Spinnaker]({{< relref "/docs/guides/continuous-delivery/spinnaker" >}}), or [one of the many available options]({{< relref "/docs/guides/continuous-delivery" >}}, so that you can deploy continuously as you merge code. For even more advanced scenarios, we might want to build a custom program such as a CLI that uses the [Automation API]({{< relref "/blog/automation-api" >}}) to perform these actions behind a simpler, purpose-built interface.
```